### ISSUE TITLE
Why does docker write rules to iptables repeatedly

### ISSUE THREAD

**Issue by FOGBABY** [moby/moby#41286]
Hi

  When I started docker, I found that docker was repeatedly writing some rules to iptables. For example,execute the following command: 
                                                                                       docker run -p 80:80 
  iptables in the host will add the following new rules:
                -A DOCKER -d 172.18.0.4/32 ! -i br-507a9fd6942b -o br-507a9fd6942b -p tcp -m tcp --dport 80 -j ACCEPT 
  At this point I re-execute the docker run command，found that iptables added a duplicate rule:
                 -A DOCKER -d 172.18.0.4/32 ! -i br-507a9fd6942b -o br-507a9fd6942b -p tcp -m tcp --dport 80 -j ACCEPT 
 This means that I have to delete the port mapping rule twice to ensure that the deletion is successful. So how to solve the above problems by configuring docker to ensure that docker will not add rules repeatedly？

I would be grateful for any advice

**Comment by thaJeztah** [comment]
Please provide more information as requested in the [issue-template](https://raw.githubusercontent.com/docker/docker/master/.github/ISSUE_TEMPLATE.md) that was shown when you opened this issue.

- provide the output of `docker version`
- provide the output of `docker info`
- provide the exact steps to reproduce. If the steps to reproduce require custom images, or source code, provide a _minimal_ test-case to reproduce. Keep in mind that we don't have access to your private images or source code. Without steps to reproduce, or a reproducible case it's unlikely that your issue can be resolved.


I tried reproducing what you described on an Ubuntu 20.04 machine with Docker 19.03.12, but I don't see the issue happening. It's possible there's a bug, but without exact steps to reproduce, it may be hard to look into that.

Here's what I tried;

Store the iptables rules in `before.txt`

```bash
iptables -L -v -n > before.txt
```

Start a container that publishes port 80, and exit the container after it started

```bash
docker run -p 80:80 nginx:alpine
^C
```

Store the iptables rules, and compare with "before"

```bash
iptables -L -v -n > after1.txt

git diff --no-index before.txt after1.txt
```

```patch
diff --git a/before.txt b/after1.txt
index a233521..9ddbfd6 100644
--- a/before.txt
+++ b/after1.txt
@@ -1,4 +1,4 @@
-Chain INPUT (policy ACCEPT 950 packets, 876K bytes)
+Chain INPUT (policy ACCEPT 41 packets, 2828 bytes)
  pkts bytes target     prot opt in     out     source               destination

 Chain FORWARD (policy DROP 0 packets, 0 bytes)
@@ -10,7 +10,7 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
     0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
     0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0

-Chain OUTPUT (policy ACCEPT 628 packets, 74064 bytes)
+Chain OUTPUT (policy ACCEPT 22 packets, 2192 bytes)
  pkts bytes target     prot opt in     out     source               destination

 Chain DOCKER (1 references)
```

Start a container in the background that publishes port 80, store the iptables rules while it's running, and compare with the previous state;

```bash
docker run -d --name mycontainer -p 80:80 nginx:alpine

iptables -L -v -n > after2.txt

git diff --no-index after1.txt after2.txt
```

```patch
iff --git a/after1.txt b/after2.txt
index 9ddbfd6..c86fe74 100644
--- a/after1.txt
+++ b/after2.txt
@@ -1,4 +1,4 @@
-Chain INPUT (policy ACCEPT 41 packets, 2828 bytes)
+Chain INPUT (policy ACCEPT 5 packets, 356 bytes)
  pkts bytes target     prot opt in     out     source               destination

 Chain FORWARD (policy DROP 0 packets, 0 bytes)
@@ -10,11 +10,12 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
     0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
     0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0

-Chain OUTPUT (policy ACCEPT 22 packets, 2192 bytes)
+Chain OUTPUT (policy ACCEPT 4 packets, 448 bytes)
  pkts bytes target     prot opt in     out     source               destination

 Chain DOCKER (1 references)
  pkts bytes target     prot opt in     out     source               destination
+    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.2           tcp dpt:80

 Chain DOCKER-ISOLATION-STAGE-1 (1 references)
  pkts bytes target     prot opt in     out     source               destination
```

Stop the container, store the iptables rules, and compare with the previous state;

```bash
docker stop mycontainer

iptables -L -v -n > after3.txt

git diff --no-index after2.txt after3.txt
```

```patch
diff --git a/after2.txt b/after3.txt
index c86fe74..d8f8260 100644
--- a/after2.txt
+++ b/after3.txt
@@ -1,4 +1,4 @@
-Chain INPUT (policy ACCEPT 5 packets, 356 bytes)
+Chain INPUT (policy ACCEPT 5 packets, 364 bytes)
  pkts bytes target     prot opt in     out     source               destination

 Chain FORWARD (policy DROP 0 packets, 0 bytes)
@@ -10,12 +10,11 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
     0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
     0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0

-Chain OUTPUT (policy ACCEPT 4 packets, 448 bytes)
+Chain OUTPUT (policy ACCEPT 4 packets, 512 bytes)
  pkts bytes target     prot opt in     out     source               destination

 Chain DOCKER (1 references)
  pkts bytes target     prot opt in     out     source               destination
-    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.2           tcp dpt:80

 Chain DOCKER-ISOLATION-STAGE-1 (1 references)
  pkts bytes target     prot opt in     out     source               destination
```

Compare it with the initial state (`before.txt`) to verify no additional rules were left behind

```patch
root@ubuntu-s-1vcpu-2gb-ams3-01:~/repro-3976/empty# git diff --no-index after3.txt before.txt
diff --git a/after3.txt b/before.txt
index d8f8260..a233521 100644
--- a/after3.txt
+++ b/before.txt
@@ -1,4 +1,4 @@
-Chain INPUT (policy ACCEPT 5 packets, 364 bytes)
+Chain INPUT (policy ACCEPT 950 packets, 876K bytes)
  pkts bytes target     prot opt in     out     source               destination

 Chain FORWARD (policy DROP 0 packets, 0 bytes)
@@ -10,7 +10,7 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
     0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
     0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0

-Chain OUTPUT (policy ACCEPT 4 packets, 512 bytes)
+Chain OUTPUT (policy ACCEPT 628 packets, 74064 bytes)
  pkts bytes target     prot opt in     out     source               destination

 Chain DOCKER (1 references)
```

**Comment by FOGBABY** [comment]
> Please provide more information as requested in the [issue-template](https://raw.githubusercontent.com/docker/docker/master/.github/ISSUE_TEMPLATE.md) that was shown when you opened this issue.
> 
> * provide the output of `docker version`
> * provide the output of `docker info`
> * provide the exact steps to reproduce. If the steps to reproduce require custom images, or source code, provide a _minimal_ test-case to reproduce. Keep in mind that we don't have access to your private images or source code. Without steps to reproduce, or a reproducible case it's unlikely that your issue can be resolved.
> 
> I tried reproducing what you described on an Ubuntu 20.04 machine with Docker 19.03.12, but I don't see the issue happening. It's possible there's a bug, but without exact steps to reproduce, it may be hard to look into that.
> 
> Here's what I tried;
> 
> Store the iptables rules in `before.txt`
> 
> ```shell
> iptables -L -v -n > before.txt
> ```
> 
> Start a container that publishes port 80, and exit the container after it started
> 
> ```shell
> docker run -p 80:80 nginx:alpine
> ^C
> ```
> 
> Store the iptables rules, and compare with "before"
> 
> ```shell
> iptables -L -v -n > after1.txt
> 
> git diff --no-index before.txt after1.txt
> ```
> 
> ```diff
> diff --git a/before.txt b/after1.txt
> index a233521..9ddbfd6 100644
> --- a/before.txt
> +++ b/after1.txt
> @@ -1,4 +1,4 @@
> -Chain INPUT (policy ACCEPT 950 packets, 876K bytes)
> +Chain INPUT (policy ACCEPT 41 packets, 2828 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain FORWARD (policy DROP 0 packets, 0 bytes)
> @@ -10,7 +10,7 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
>      0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
>      0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
> 
> -Chain OUTPUT (policy ACCEPT 628 packets, 74064 bytes)
> +Chain OUTPUT (policy ACCEPT 22 packets, 2192 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain DOCKER (1 references)
> ```
> 
> Start a container in the background that publishes port 80, store the iptables rules while it's running, and compare with the previous state;
> 
> ```shell
> docker run -d --name mycontainer -p 80:80 nginx:alpine
> 
> iptables -L -v -n > after2.txt
> 
> git diff --no-index after1.txt after2.txt
> ```
> 
> ```diff
> iff --git a/after1.txt b/after2.txt
> index 9ddbfd6..c86fe74 100644
> --- a/after1.txt
> +++ b/after2.txt
> @@ -1,4 +1,4 @@
> -Chain INPUT (policy ACCEPT 41 packets, 2828 bytes)
> +Chain INPUT (policy ACCEPT 5 packets, 356 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain FORWARD (policy DROP 0 packets, 0 bytes)
> @@ -10,11 +10,12 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
>      0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
>      0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
> 
> -Chain OUTPUT (policy ACCEPT 22 packets, 2192 bytes)
> +Chain OUTPUT (policy ACCEPT 4 packets, 448 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain DOCKER (1 references)
>   pkts bytes target     prot opt in     out     source               destination
> +    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.2           tcp dpt:80
> 
>  Chain DOCKER-ISOLATION-STAGE-1 (1 references)
>   pkts bytes target     prot opt in     out     source               destination
> ```
> 
> Stop the container, store the iptables rules, and compare with the previous state;
> 
> ```shell
> docker stop mycontainer
> 
> iptables -L -v -n > after3.txt
> 
> git diff --no-index after2.txt after3.txt
> ```
> 
> ```diff
> diff --git a/after2.txt b/after3.txt
> index c86fe74..d8f8260 100644
> --- a/after2.txt
> +++ b/after3.txt
> @@ -1,4 +1,4 @@
> -Chain INPUT (policy ACCEPT 5 packets, 356 bytes)
> +Chain INPUT (policy ACCEPT 5 packets, 364 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain FORWARD (policy DROP 0 packets, 0 bytes)
> @@ -10,12 +10,11 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
>      0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
>      0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
> 
> -Chain OUTPUT (policy ACCEPT 4 packets, 448 bytes)
> +Chain OUTPUT (policy ACCEPT 4 packets, 512 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain DOCKER (1 references)
>   pkts bytes target     prot opt in     out     source               destination
> -    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.2           tcp dpt:80
> 
>  Chain DOCKER-ISOLATION-STAGE-1 (1 references)
>   pkts bytes target     prot opt in     out     source               destination
> ```
> 
> Compare it with the initial state (`before.txt`) to verify no additional rules were left behind
> 
> ```diff
> root@ubuntu-s-1vcpu-2gb-ams3-01:~/repro-3976/empty# git diff --no-index after3.txt before.txt
> diff --git a/after3.txt b/before.txt
> index d8f8260..a233521 100644
> --- a/after3.txt
> +++ b/before.txt
> @@ -1,4 +1,4 @@
> -Chain INPUT (policy ACCEPT 5 packets, 364 bytes)
> +Chain INPUT (policy ACCEPT 950 packets, 876K bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain FORWARD (policy DROP 0 packets, 0 bytes)
> @@ -10,7 +10,7 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
>      0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
>      0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
> 
> -Chain OUTPUT (policy ACCEPT 4 packets, 512 bytes)
> +Chain OUTPUT (policy ACCEPT 628 packets, 74064 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain DOCKER (1 references)
> ```

**Comment by FOGBABY** [comment]
> Please provide more information as requested in the [issue-template](https://raw.githubusercontent.com/docker/docker/master/.github/ISSUE_TEMPLATE.md) that was shown when you opened this issue.
> 
> * provide the output of `docker version`
> * provide the output of `docker info`
> * provide the exact steps to reproduce. If the steps to reproduce require custom images, or source code, provide a _minimal_ test-case to reproduce. Keep in mind that we don't have access to your private images or source code. Without steps to reproduce, or a reproducible case it's unlikely that your issue can be resolved.
> 
> I tried reproducing what you described on an Ubuntu 20.04 machine with Docker 19.03.12, but I don't see the issue happening. It's possible there's a bug, but without exact steps to reproduce, it may be hard to look into that.
> 
> Here's what I tried;
> 
> Store the iptables rules in `before.txt`
> 
> ```shell
> iptables -L -v -n > before.txt
> ```
> 
> Start a container that publishes port 80, and exit the container after it started
> 
> ```shell
> docker run -p 80:80 nginx:alpine
> ^C
> ```
> 
> Store the iptables rules, and compare with "before"
> 
> ```shell
> iptables -L -v -n > after1.txt
> 
> git diff --no-index before.txt after1.txt
> ```
> 
> ```diff
> diff --git a/before.txt b/after1.txt
> index a233521..9ddbfd6 100644
> --- a/before.txt
> +++ b/after1.txt
> @@ -1,4 +1,4 @@
> -Chain INPUT (policy ACCEPT 950 packets, 876K bytes)
> +Chain INPUT (policy ACCEPT 41 packets, 2828 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain FORWARD (policy DROP 0 packets, 0 bytes)
> @@ -10,7 +10,7 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
>      0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
>      0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
> 
> -Chain OUTPUT (policy ACCEPT 628 packets, 74064 bytes)
> +Chain OUTPUT (policy ACCEPT 22 packets, 2192 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain DOCKER (1 references)
> ```
> 
> Start a container in the background that publishes port 80, store the iptables rules while it's running, and compare with the previous state;
> 
> ```shell
> docker run -d --name mycontainer -p 80:80 nginx:alpine
> 
> iptables -L -v -n > after2.txt
> 
> git diff --no-index after1.txt after2.txt
> ```
> 
> ```diff
> iff --git a/after1.txt b/after2.txt
> index 9ddbfd6..c86fe74 100644
> --- a/after1.txt
> +++ b/after2.txt
> @@ -1,4 +1,4 @@
> -Chain INPUT (policy ACCEPT 41 packets, 2828 bytes)
> +Chain INPUT (policy ACCEPT 5 packets, 356 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain FORWARD (policy DROP 0 packets, 0 bytes)
> @@ -10,11 +10,12 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
>      0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
>      0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
> 
> -Chain OUTPUT (policy ACCEPT 22 packets, 2192 bytes)
> +Chain OUTPUT (policy ACCEPT 4 packets, 448 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain DOCKER (1 references)
>   pkts bytes target     prot opt in     out     source               destination
> +    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.2           tcp dpt:80
> 
>  Chain DOCKER-ISOLATION-STAGE-1 (1 references)
>   pkts bytes target     prot opt in     out     source               destination
> ```
> 
> Stop the container, store the iptables rules, and compare with the previous state;
> 
> ```shell
> docker stop mycontainer
> 
> iptables -L -v -n > after3.txt
> 
> git diff --no-index after2.txt after3.txt
> ```
> 
> ```diff
> diff --git a/after2.txt b/after3.txt
> index c86fe74..d8f8260 100644
> --- a/after2.txt
> +++ b/after3.txt
> @@ -1,4 +1,4 @@
> -Chain INPUT (policy ACCEPT 5 packets, 356 bytes)
> +Chain INPUT (policy ACCEPT 5 packets, 364 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain FORWARD (policy DROP 0 packets, 0 bytes)
> @@ -10,12 +10,11 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
>      0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
>      0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
> 
> -Chain OUTPUT (policy ACCEPT 4 packets, 448 bytes)
> +Chain OUTPUT (policy ACCEPT 4 packets, 512 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain DOCKER (1 references)
>   pkts bytes target     prot opt in     out     source               destination
> -    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.2           tcp dpt:80
> 
>  Chain DOCKER-ISOLATION-STAGE-1 (1 references)
>   pkts bytes target     prot opt in     out     source               destination
> ```
> 
> Compare it with the initial state (`before.txt`) to verify no additional rules were left behind
> 
> ```diff
> root@ubuntu-s-1vcpu-2gb-ams3-01:~/repro-3976/empty# git diff --no-index after3.txt before.txt
> diff --git a/after3.txt b/before.txt
> index d8f8260..a233521 100644
> --- a/after3.txt
> +++ b/before.txt
> @@ -1,4 +1,4 @@
> -Chain INPUT (policy ACCEPT 5 packets, 364 bytes)
> +Chain INPUT (policy ACCEPT 950 packets, 876K bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain FORWARD (policy DROP 0 packets, 0 bytes)
> @@ -10,7 +10,7 @@ Chain FORWARD (policy DROP 0 packets, 0 bytes)
>      0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
>      0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
> 
> -Chain OUTPUT (policy ACCEPT 4 packets, 512 bytes)
> +Chain OUTPUT (policy ACCEPT 628 packets, 74064 bytes)
>   pkts bytes target     prot opt in     out     source               destination
> 
>  Chain DOCKER (1 references)
> ```

Thanks for your reply.
out of docker info:
> ```shell
> Containers: 2
> Running: 0
> Paused: 0
> Stopped: 2
>Images: 15
>Server Version: 1.13.1
>Storage Driver: devicemapper
> Pool Name: docker-8:2-5637513-pool
> Pool Blocksize: 65.54 kB
> Base Device Size: 10.74 GB
> Backing Filesystem: xfs
> Data file: /dev/loop0
> Metadata file: /dev/loop1
> Data Space Used: 3.18 GB
> Data Space Total: 107.4 GB
> Data Space Available: 104.2 GB
> Metadata Space Used: 5.808 MB
> Metadata Space Total: 2.147 GB
> Metadata Space Available: 2.142 GB
> Thin Pool Minimum Free Space: 10.74 GB
> Udev Sync Supported: true
> Deferred Removal Enabled: false
> Deferred Deletion Enabled: false
> Deferred Deleted Device Count: 0
> Data loop file: /var/lib/docker/devicemapper/devicemapper/data
> WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
> Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
> Library Version: 1.02.117-RHEL6 (2016-12-13)
>Logging Driver: json-file
>Cgroup Driver: cgroupfs
>Plugins:
> Volume: local
> Network: bridge host macvlan null overlay
>Swarm: inactive
>Runtimes: runc
>Default Runtime: runc
>Init Binary: docker-init
>containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
>runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
>init version: 949e6fa
>Kernel Version: 4.4.12-rt19.ns6
>Operating System: <unknown>
>OSType: linux
>Architecture: x86_64
>CPUs: 2
>Total Memory: 3.861 GiB
>Name: test
>ID: H5CC:LUE4:YZXK:XWLG:LG2J:U4Y6:RJOE:SJLW:PRUR:EUSE:UDZR:QPJX
>Docker Root Dir: /var/lib/docker
>Debug Mode (client): false
>Debug Mode (server): false
>Registry: https://index.docker.io/v1/
>Experimental: false
>Insecure Registries:
> 127.0.0.0/8
>Live Restore Enabled: false
> ^C

out of docker version:
> ```shell
>  Version:      1.13.1

First of all,I try to run the following command：
> ```shell
>  docker run --name=mysql --restart=always --privileged=true --network=test --ip=172.18.0.4 -p 3306:3306 -v /etc/localtime:/etc/localtime -v /home/fogbaby/db/cnf:/etc/mysql/conf.d -v /home/fogbaby/db/data:/var/lib/mysql -v /home/fogbaby/db/logs:/var/log/mysql -itd mysql
>   ^C

And at this time, I try to check the status of Iptables：
> ```shell
> service iptables status

The status of Iptables is as follows:
> ```shell
>  Chain PREROUTING (policy ACCEPT)
>  num  target     prot opt source               destination
> 1    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> 2    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> 3    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> Chain INPUT (policy ACCEPT)
> num  target     prot opt source               destination
> Chain OUTPUT (policy ACCEPT)
> num  target     prot opt source               destination
> 1    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> 2    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> 3    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> Chain POSTROUTING (policy ACCEPT)
> num  target     prot opt source               destination
> 1    MASQUERADE  all  --  172.17.0.0/16        0.0.0.0/0
> 2    MASQUERADE  all  --  172.17.0.0/16        0.0.0.0/0
> 3    MASQUERADE  all  --  172.18.0.0/16        0.0.0.0/0
> 4    MASQUERADE  tcp  --  172.18.0.4           172.18.0.4          tcp dpt:3306
> Chain DOCKER (3 references)
> num  target     prot opt source               destination
> 1    ACCEPT     tcp  --  0.0.0.0/0            172.18.0.4          tcp dpt:3306
> Chain DOCKER-ISOLATION (3 references)
> num  target     prot opt source               destination
> 1    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 2    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 3    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 4    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 5    RETURN     all  --  0.0.0.0/0            0.0.0.0/0
> Chain DOCKER-ISOLATION (3 references)
> num  target     prot opt source               destination
> 1    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 2    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 3    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 4    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 5    RETURN     all  --  0.0.0.0/0            0.0.0.0/0
> Chain PREROUTING (policy ACCEPT)
> num  target     prot opt source               destination
> 1    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> 2    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> 3    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> Chain INPUT (policy ACCEPT)
> num  target     prot opt source               destination
> Chain OUTPUT (policy ACCEPT)
> num  target     prot opt source               destination
> 1    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> 2    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> 3    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> Chain POSTROUTING (policy ACCEPT)
> num  target     prot opt source               destination
> 1    MASQUERADE  all  --  172.17.0.0/16        0.0.0.0/0
> 2    MASQUERADE  all  --  172.18.0.0/16        0.0.0.0/0
> 3    MASQUERADE  all  --  172.17.0.0/16        0.0.0.0/0
> 4    MASQUERADE  tcp  --  172.18.0.4           172.18.0.4          tcp dpt:3306
> Chain DOCKER (6 references)
> num  target     prot opt source               destination
> 1    RETURN     all  --  0.0.0.0/0            0.0.0.0/0
> 2    RETURN     all  --  0.0.0.0/0            0.0.0.0/0
> 3    RETURN     all  --  0.0.0.0/0            0.0.0.0/0
> 4    DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0           tcp dpt:3306 to:172.18.0.4:3306

At this point I re-execute the docker run command，found that iptables  added a duplicate rule：
> ```shell
>  Chain PREROUTING (policy ACCEPT)
>  num  target     prot opt source               destination
> 1    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> 2    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> 3    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> Chain INPUT (policy ACCEPT)
> num  target     prot opt source               destination
> Chain OUTPUT (policy ACCEPT)
> num  target     prot opt source               destination
> 1    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> 2    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> 3    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> Chain POSTROUTING (policy ACCEPT)
> num  target     prot opt source               destination
> 1    MASQUERADE  all  --  172.17.0.0/16        0.0.0.0/0
> 2    MASQUERADE  all  --  172.17.0.0/16        0.0.0.0/0
> 3    MASQUERADE  all  --  172.18.0.0/16        0.0.0.0/0
> 4    MASQUERADE  tcp  --  172.18.0.4           172.18.0.4          tcp dpt:3306
> 5    MASQUERADE  tcp  --  172.18.0.4           172.18.0.4          tcp dpt:3306
> Chain DOCKER (3 references)
> num  target     prot opt source               destination
> 1    ACCEPT     tcp  --  0.0.0.0/0            172.18.0.4          tcp dpt:3306
> 2    ACCEPT     tcp  --  0.0.0.0/0            172.18.0.4          tcp dpt:3306
> Chain DOCKER-ISOLATION (3 references)
> num  target     prot opt source               destination
> 1    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 2    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 3    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 4    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 5    RETURN     all  --  0.0.0.0/0            0.0.0.0/0
> Chain DOCKER-ISOLATION (3 references)
> num  target     prot opt source               destination
> 1    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 2    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 3    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 4    DROP       all  --  0.0.0.0/0            0.0.0.0/0
> 5    RETURN     all  --  0.0.0.0/0            0.0.0.0/0
> Chain PREROUTING (policy ACCEPT)
> num  target     prot opt source               destination
> 1    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> 2    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> 3    DOCKER     all  --  0.0.0.0/0            0.0.0.0/0           ADDRTYPE match dst-type LOCAL
> Chain INPUT (policy ACCEPT)
> num  target     prot opt source               destination
> Chain OUTPUT (policy ACCEPT)
> num  target     prot opt source               destination
> 1    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> 2    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> 3    DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8         ADDRTYPE match dst-type LOCAL
> Chain POSTROUTING (policy ACCEPT)
> num  target     prot opt source               destination
> 1    MASQUERADE  all  --  172.17.0.0/16        0.0.0.0/0
> 2    MASQUERADE  all  --  172.18.0.0/16        0.0.0.0/0
> 3    MASQUERADE  all  --  172.17.0.0/16        0.0.0.0/0
> 4    MASQUERADE  tcp  --  172.18.0.4           172.18.0.4          tcp dpt:3306
> 5    MASQUERADE  tcp  --  172.18.0.4           172.18.0.4          tcp dpt:3306
> Chain DOCKER (6 references)
> num  target     prot opt source               destination
> 1    RETURN     all  --  0.0.0.0/0            0.0.0.0/0
> 2    RETURN     all  --  0.0.0.0/0            0.0.0.0/0
> 3    RETURN     all  --  0.0.0.0/0            0.0.0.0/0
> 4    DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0           tcp dpt:3306 to:172.18.0.4:3306
> 5    DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0           tcp dpt:3306 to:172.18.0.4:3306

This means that if I want to delete port 3306 without restarting Iptables, I need to delete it twice to succeed. 

Hope the above information can make you understand more clearly.

**Comment by thaJeztah** [comment]
Docker 1.13.1 is really old (reached EOL 3 years ago, has known vulnerabilities and has not been maintained since) and, from the output of `docker info`, I suspect you're running the Red Hat fork of Docker, which has a lot of modifications that are not in this repository (and which are known to cause problems). I also see you seem to be running RHEL 6, which is no longer supported for Docker (or by Red Hat for running containers), with a non-standard kernel for that version of Red Hat.


I don't think we'll be able to help with this issue, and it's possible the issue has been fixed in newer versions, but feel free to report if you're seeing the same problem on a currently maintained version.
